## Proyecto Final Interconnect
Al operador de telecomunicaciones Interconnect le gustaría poder pronosticar su tasa de cancelación de clientes. Si se descubre que un usuario o usuaria planea irse, se le ofrecerán códigos promocionales y opciones de planes especiales. El equipo de marketing de Interconnect ha recopilado algunos de los datos personales de sus clientes, incluyendo información sobre sus planes y contratos.

Lo primero a realizar en esta parte del proyecto es realizar un EDA de los datos disponibles para el trabajo. Finalmente responder a las preguntas que nosotros mismos realizaremos como, cuantos clientes se fueron, cuantos clientes se quedaron, que datos nulos, o sucios, hay en el dataset. Ver si hay alguna irregularidad o algo por el estilo.

In [3]:
import numpy as np
import pandas as pd

#Primero cargaremos los 4 datasets de la carpeta de final_provider.

contract_data=pd.read_csv('final_provider/contract.csv')
personal_data=pd.read_csv('final_provider/personal.csv')
internet_data=pd.read_csv('final_provider/internet.csv')
phone_data=pd.read_csv('final_provider/phone.csv')

In [4]:
# Exploramos las dimensiones de cada dataset
print("Contract data shape:", contract_data.shape)
print("Personal data shape:", personal_data.shape) 
print("Internet data shape:", internet_data.shape)
print("Phone data shape:", phone_data.shape)

Contract data shape: (7043, 8)
Personal data shape: (7043, 5)
Internet data shape: (5517, 8)
Phone data shape: (6361, 2)


Algo curioso que podemos ver es que todos tienen distintas cantidades de filas, lo que significa que puede ser que haya mas clientes con servicios de internet o de telefono o de diferentes cosas.

In [5]:
#Empezamos el EDA con el primer dataset de contract
print('==Contract Data==')
display(contract_data.head(10))
display('\n',contract_data.info())

==Contract Data==


,customerID,BeginDate,EndDate,Type,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,7590-VHVEG,2020-01-01,No,Month-to-month,Yes,Electronic check,29.85,29.85
1,5575-GNVDE,2017-04-01,No,One year,No,Mailed check,56.95,1889.5
2,3668-QPYBK,2019-10-01,2019-12-01 00:00:00,Month-to-month,Yes,Mailed check,53.85,108.15
3,7795-CFOCW,2016-05-01,No,One year,No,Bank transfer (automatic),42.30,1840.75
4,9237-HQITU,2019-09-01,2019-11-01 00:00:00,Month-to-month,Yes,Electronic check,70.70,151.65
5,9305-CDSKC,2019-03-01,2019-11-01 00:00:00,Month-to-month,Yes,Electronic check,99.65,820.5
6,1452-KIOVK,2018-04-01,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.4
7,6713-OKOMC,2019-04-01,No,Month-to-month,No,Mailed check,29.75,301.9
8,7892-POOKP,2017-07-01,2019-11-01 00:00:00,Month-to-month,Yes,Electronic check,104.80,3046.05
9,6388-TABGU,2014-12-01,No,One year,No,Bank transfer (automatic),56.15,3487.95


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


'\n'

None

Para este dataset, encontramos 0 datos nulos, encontramos 7043 contratos, lo que significa que es el numero total de clientes que hay. Encontramos en este dataset, sus customerID que es bastante importante pues luego podriamos juntar datasets en base a sus customerID. Tambien tenemos sus fechas de inicio de contrato y sus fechas de finalizacion de contrato si es que tienen, tenemos cuanto pagan y como pagan.

In [6]:
# Ver los valores únicos en EndDate
print("Valores únicos en EndDate:")
print(contract_data['EndDate'].value_counts())

Valores únicos en EndDate:
EndDate
No                     5174
2019-11-01 00:00:00     485
2019-12-01 00:00:00     466
2020-01-01 00:00:00     460
2019-10-01 00:00:00     458
Name: count, dtype: int64


Con esto podemos responder que hay 5174 clientes que no dieron de baja los servicios, podemos encontrar que dan de baja el servicio el dia primero de cada mes, y vemos que el promedio de estos clientes es de 450 a 500 clientes.

In [7]:
#Continuamos con el dataframe de Personal.
print('==Personal data==')
display(personal_data.head())
display('\n',personal_data.info())

==Personal data==


,customerID,gender,SeniorCitizen,Partner,Dependents
0,7590-VHVEG,Female,0,Yes,No
1,5575-GNVDE,Male,0,No,No
2,3668-QPYBK,Male,0,No,No
3,7795-CFOCW,Male,0,No,No
4,9237-HQITU,Female,0,No,No


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


'\n'

None

En este dataframe la informacion es bastante sencilla, describe que genero son los clientes, si tienen parejas o dependientes y si son adultos o no.

In [8]:
#Continuamos con el dataframe de Internet.
print('==Internet data==')
display(internet_data.head())
display('\n',internet_data.info())

==Internet data==


,customerID,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies
0,7590-VHVEG,DSL,No,Yes,No,No,No,No
1,5575-GNVDE,DSL,Yes,No,Yes,No,No,No
2,3668-QPYBK,DSL,Yes,Yes,No,No,No,No
3,7795-CFOCW,DSL,Yes,No,Yes,Yes,No,No
4,9237-HQITU,Fiber optic,No,No,No,No,No,No


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB


'\n'

None

En este dataframe se encuentra el segundo tipo de servicio es deci, Internet.}
Este servicio ofrece una red que se puede configurar a través de una línea telefónica es decir DSL, o a través de un cable de fibra óptica

En este servicio de internet podemos ver que ofrecen otros servicios extras, seguridad en internet, soporte tecnico, almacenamiento en la nube, streaming de tv o streaming de peliculas.

In [9]:
#Finalmente sigue el de phone data.
print('==Phone data==')
display(phone_data.head())
display('\n',phone_data.info())

==Phone data==


,customerID,MultipleLines
0,5575-GNVDE,No
1,3668-QPYBK,No
2,9237-HQITU,No
3,9305-CDSKC,Yes
4,1452-KIOVK,Yes


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


'\n'

None

En este primer tipo de servicio se ofrece: Comunicación por teléfono fijo. El teléfono se puede conectar a varias líneas de manera simultánea.

# Preguntas aclaratorias.

Fecha de Corte del Dataset: ¿Cuál es la fecha exacta en la que se realizó la extracción de los datos? Esto es fundamental para calcular con precisión la antigüedad de los clientes que aún aparecen como "No" en EndDate.

Criterio de Corte de Servicio: ¿Se considera "baja" en el momento exacto de la cancelación, o existe un periodo de gracia (ej. 30 días) antes de marcar al cliente como perdido?

Métrica de Éxito de Negocio: Aunque el objetivo técnico es el AUC-ROC, ¿el negocio prefiere priorizar el Recall (identificar a la mayor cantidad posible de desertores, aunque haya falsas alarmas) o la Precisión (estar muy seguros antes de ofrecer una promoción de retención)?

Servicios Adicionales: ¿Hay algún servicio de los adicionales que el operador de telecomunicaciones Interconnect considere mas importante de retención por encima de los demás?

# Plan de Trabajo: Modelo de Predicción de Cortes de Servicio.

### Fase 1: Unificación de Datos

La primera parte que realizaremos sera realizar un left join usando la de contratos como base, y unir todos los dataframes en una mediante el CustomerID

Los clientes que no tengan internet o telefono se les marcaran con un NO o un NA o algo por el estilo, sin borrar estos datos, ya que es informacion valiosa.

Habria que convertir las fechas a datetime y los cargos totales a numericos si es que no lo estan. Para poder manejar estos datos de mejor manera e incluso agregar columnas como cuantos dias tuvieron de servicio totales, o cosas por el estilo, para el analisis de datos.

### Fase 2: Manejo de Columnas.

Crearemos la columna target, en base a la columna de EndDate, tomaremos 0 si EndDate= No, y 1 si corto el servicio.

Calcularemos los dias totales como clientes, ya que la Lealtad de los clientes es un valor que importa mucho.

Aplicaremos One-Hot Encodign para las variables categoricas.

### Fase 3: Entrenamiento del Modelo

Empezaremos con la division de datos para el modelo, en este caso, lo dividiremos en 60/20/20 para entrenamiento validacion y prueba, a menos que se especifique algo distinto mas adelante.

Ya que hay muchos clientes mas activos que no activos, revisaremos si habra que equilibar las clases con tecnicas como upsampling o algo por el estilo, si no con el peso de las clases del modelo.

Probaremos un modelo base, para despues avanzar con un modelo de Gradient Boosting.

### Fase 4: Evaluación y Ajuste

Para la evaluacion tomaremos el AUC-ROC y el Accuracy, es mas importante no perder los clientes sin antes ofrecerles promociones que desconecten el servicio sin haber sabido que lo harian. Es decir en este caso se valorara mas que encontremos a todos los casos, que la exactitud de si lo cortaran o no.
